### ISSUE TITLE
Expression template containing double-quoted string fails

### ISSUE THREAD

**Issue by mac9416** [argoproj/argo-workflows#6260]
### tl;dr

Argo failed to evaluate an expression template containing a double-quoted string.

```yaml
value: '{{=jsonpath(workflow.parameters.config, "$.a")}}'
```

But it successfully evaluated an expression template containing a single-quoted string.

```yaml
value: "{{=jsonpath(workflow.parameters.config, '$.a')}}"
```

According to the [expr language definition](https://github.com/antonmedv/expr/blob/da4c23e1ae39c67a7e2f5bc96d772573483ec359/docs/Language-Definition.md), it supports either.

#### Fails

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: expression-flatten-json-
spec:
  arguments:
    parameters:
      - name: config
        value: '{"a": "1"}'
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: a
            value: '{{=jsonpath(workflow.parameters.config, "$.a")}}'
      script:
        env:
          - name: A
            value: "{{inputs.parameters.a}}"
        image: debian:9.4
        command: [bash]
        source: |
          echo "$A"
```

Output is:

```
{{=jsonpath(workflow.parameters.config, "$.a")}}
```

#### Succeeds

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: expression-flatten-json-
spec:
  arguments:
    parameters:
      - name: config
        value: '{"a-hyphenated": "1"}'
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: a
            value: "{{=jsonpath(workflow.parameters.config, '$.a')}}"
      script:
        env:
          - name: A
            value: "{{inputs.parameters.a}}"
        image: debian:9.4
        command: [bash]
        source: |
          echo "$A"
```

Output is:

```
1
```

### Why it matters

There's no easy/intuitive way to use a string that contains single-quotes. I would expect to be able to use `{{=jsonpath(workflow.parameters.config, "$['a-hyphenated']")}}`, but that's not possible with double-quoted strings off the table.

I have also tried escaping the inner quotes, but that does not work either. I also don't see any documentation in expr for escaping characters. (Escaped once for YAML, another time for expr.)

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: expression-flatten-json-
spec:
  arguments:
    parameters:
      - name: config
        value: '{"a-hyphenated": "1"}'
  entrypoint: main
  templates:
    - name: main
      inputs:
        parameters:
          - name: a
            value: "{{=jsonpath(workflow.parameters.config, '$[\\\'a-hyphenated\\\']')}}"
      script:
        env:
          - name: A
            value: "{{inputs.parameters.a}}"
        image: debian:9.4
        command: [bash]
        source: |
          echo "$A"
```

## Diagnostics

Kubernetes provider: Docker Desktop

Argo version: 3.1.0

Executor: The default from install.yaml... I assume it's Docker.

It's not a regression, since expression templates is a new feature.

Controller logs from failed workflow example above:

```
$ kubectl logs -n argo deploy/workflow-controller | grep expression-flatten-json-xc7x8
time="2021-06-30T14:24:14.736Z" level=info msg="Processing workflow" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:14.745Z" level=info msg="Updated phase  -> Running" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:14.748Z" level=info msg="Pod node expression-flatten-json-xc7x8 initialized Pending" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:14.774Z" level=info msg="Created pod: expression-flatten-json-xc7x8 (expression-flatten-json-xc7x8)" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:14.790Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=8924265 workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Processing workflow" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Updating node expression-flatten-json-xc7x8 exit code 0" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Updating node expression-flatten-json-xc7x8 status Pending -> Succeeded" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Updated phase Running -> Succeeded" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Marking workflow completed" namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.778Z" level=info msg="Checking daemoned children of " namespace=default workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.788Z" level=info msg="Workflow update successful" namespace=default phase=Succeeded resourceVersion=8924295 workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:24.794Z" level=info msg="cleaning up pod" action=labelPodCompleted key=default/expression-flatten-json-xc7x8/labelPodCompleted
```

wait container logs from failed workflow example above:

```
$ kubectl logs -c wait -l workflows.argoproj.io/workflow=expression-flatten-json-xc7x8
time="2021-06-30T14:24:19.141Z" level=info msg="listed containers" containers="map[init:{f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6 Exited {0 63760659856 <nil>}} main:{2c1238701d5cd78ea80b74b5a62f46f7591244a298c2a98c1625b207fa0e8d9d Exited {0 63760659858 <nil>}} wait:{4774fdca670b0f78a22a0d7626cb3cd6257796a364d3b4a1124bf8ac17128c41 Up {0 63760659857 <nil>}}]"
time="2021-06-30T14:24:19.141Z" level=info msg="Killing sidecars [\"init\"]"
time="2021-06-30T14:24:19.149Z" level=info msg="Get pods 200"
time="2021-06-30T14:24:19.159Z" level=info msg="docker kill --signal TERM f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6"
time="2021-06-30T14:24:19.160Z" level=info msg="listed containers" containers="map[init:{f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6 Exited {0 63760659856 <nil>}} main:{2c1238701d5cd78ea80b74b5a62f46f7591244a298c2a98c1625b207fa0e8d9d Exited {0 63760659858 <nil>}} wait:{4774fdca670b0f78a22a0d7626cb3cd6257796a364d3b4a1124bf8ac17128c41 Up {0 63760659857 <nil>}}]"
time="2021-06-30T14:24:19.183Z" level=error msg="`docker kill --signal TERM f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6` failed: Error response from daemon: Cannot kill container: f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6: Container f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6 is not running\n"
time="2021-06-30T14:24:19.183Z" level=warning msg="Ignored error from 'docker kill --signal TERM': Error response from daemon: Cannot kill container: f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6: Container f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6 is not running"
time="2021-06-30T14:24:19.183Z" level=info msg="[docker wait f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6]"
time="2021-06-30T14:24:19.206Z" level=info msg="Containers [f2701eb2f7fadd55bdd969621e97b1f0a6baa81622f73143c56185aefbb271b6] killed successfully"
time="2021-06-30T14:24:19.206Z" level=info msg="Alloc=4977 TotalAlloc=10439 Sys=73297 NumGC=4 Goroutines=10"
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by mac9416** [comment]
Here is an even simpler test:

```yaml
value: '{{="test"}}'  # evaluates to `{{="test"}}`
value: "{{='test'}}"  # evaluates to `test`
```

**Comment by sarabala1979** [comment]
would you like to submit PR this?

**Comment by mac9416** [comment]
@sarabala1979 maybe. I ran the controller locally and got some obtuse error back from expr (something about the backslash character being invalid, even when it wasn't present???). Will post that when I get the chance. If the problem turns out to be simple, I'll be happy to contribute a fix.

If this proves more involved, I might just have to submit a PR documenting the bug.

**Comment by mac9416** [comment]
FastTemplate is escaping double-quotes in the tags before they're fed to expr. I believe adding this line before evaluating the expression would solve the issue:

```go
s = strings.Replace(s, "\\\"", "\"", -1)
```

I'll try to put up a PR, but might not find the time. At least now there's a clearer idea what the issue is. :-)

**Comment by mac9416** [comment]
1. the template is [marshalled to JSON](https://github.com/argoproj/argo-workflows/blob/c9246d3d4c162e0f7fe76f2ee37c55bdbfa4b0c6/workflow/common/util.go#L171) in the `SubstituteParams` function (i.e. the interface becomes JSON as a JSON-encoded string [i.e. double quotes are escaped])
2. this string is passed to the [`Replace` function](https://github.com/argoproj/argo-workflows/blob/c9246d3d4c162e0f7fe76f2ee37c55bdbfa4b0c6/util/template/replace.go#L3)
3. which is then passed to [`NewTemplate`](https://github.com/argoproj/argo-workflows/blob/c9246d3d4c162e0f7fe76f2ee37c55bdbfa4b0c6/util/template/template.go#L22) which calls fasttemplate's `NewTemplate`
4. which [calls `Reset`](https://github.com/valyala/fasttemplate/blob/2275c48ed3e40f01fa347e220be31070a89b6ad4/template.go#L197)
5. which dutifully retrieves as a `tag` everything between `{{` and `}}`, which includes the JSON-escaped double quotes

It would be safe and sufficient to do a string replace before evaluating the expression. I'll put up a PR, but it might be a few days. Gotta finally set up a decent dev environment. :-P
